File: Medium/8/8.py
# 32-bit signed integer limits
INT_MIN = -2147483648
INT_MAX = 2147483647
DIGITS = {str(x) for x in range(10)}


class Solution:
    def myAtoi(self, s: str) -> int:
        num = 0
        negative = False
        started = False
        for c in s:
            if c == " " and not started:
                continue
            if c in DIGITS:
                started = True
                digit = int(c) if not negative else -int(c)
                if num > INT_MAX / 10 or (num == int(INT_MAX / 10) and digit > 7):
                    return INT_MAX
                if num < INT_MIN / 10 or (num == int(INT_MIN / 10) and digit < -8):
                    return INT_MIN
                num = num * 10 + digit
            elif c == "+" and not started:
                negative = False
                started = True
            elif c == "-" and not started:
                negative = True
                started = True
            else:
                break
        return num


print(f"Solution: {Solution().myAtoi("123")}")
print(f"Solution: {Solution().myAtoi("-042")}")
print(f"Solution: {Solution().myAtoi("1337c0d3")}")
print(f"Solution: {Solution().myAtoi("2147483648")}")

File: medium/8/8.js
// 32-bit signed integer limits
INT_MIN = -2147483648;
INT_MAX = 2147483647;
POS_LIM = Math.trunc(INT_MAX / 10);
NEG_LIM = Math.trunc(INT_MIN / 10);
const DIGITS = new Set(
  Array(10)
    .fill(0)
    .map((_, i) => String(i)),
);

/**
 * @param {string} s
 * @return {number}
 */
var myAtoi = function (s) {
  let num = 0;
  let started = false;
  let negative = false;
  for (c of s) {
    if (c == " " && !started) continue;
    if (DIGITS.has(c)) {
      started = true;
      let digit = !negative ? Number(c) : -Number(c);
      if (num > POS_LIM || (num == POS_LIM && digit > 7)) return INT_MAX;
      if (num < NEG_LIM || (num == NEG_LIM && digit < -8)) return INT_MIN;
      num = num * 10 + digit;
    } else if (c == "-" && !started) {
      negative = true;
      started = true;
    } else if (c == "+" && !started) {
      started = true;
    } else return num;
  }
  return num;
};

console.log("It's Showtime Folks!");
console.log(`Solution: ${myAtoi("123")}`);
console.log(`Solution: ${myAtoi("-042")}`);
console.log(`Solution: ${myAtoi("1337c0d3")}`);
console.log(`Solution: ${myAtoi("2147483648")}`);
